python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+
+    def add_object_id(self, object_id):
+        self._new_object_ids.put(object_id)
+
+    def run(self):
+        unready = copy.copy(self._object_ids)
+        while self._num_ready < self._total_object_ids:
+            # Get as many new IDs from the queue as possible without blocking,
+            # unless we have no IDs to wait on, in which case we block.
+            while True:
+                try:
+                    block = len(unready) == 0
+                    new_object_id = self._new_object_ids.get(block=block)
+                    self._add_object_id(new_object_id)
+                    unready.append(new_object_id)
+                except queue.Empty:
+                    break
+
+            ready, unready = ray.wait(unready, num_returns=1)
+            assert len(ready) == 1
+            ready_id = ready[0]
+
+            batch = ray.get(ready_id)
+            for result in batch:
+                if isinstance(result, Exception):
+                    with self._lock:
+                        self._got_error = True
+                    if self._error_callback is not None:
+                        self._error_callback(result)
+                elif self._callback is not None:
+                    self._callback(result)
+
+            with self._lock:
+                self._num_ready += 1
+                self._results[self._indices[ready_id]] = batch
+                self._ready_index_queue.put(self._indices[ready_id])
+
+        with self._lock:
+            self._done = True
+
+    def done(self):
+        with self._lock:
+            return self._done
+
+    def got_error(self):
+        with self._lock:
+            return self._got_error
+
+    def results(self):
+        with self._lock:
+            return self._results
+
+    def next_ready_index(self, timeout=None):
+        return self._ready_index_queue.get(timeout=timeout)
+
+
+class AsyncResult(object):
+    def __init__(self,
+                 chunk_object_ids,
+                 callback=None,
+                 error_callback=None,
+                 single_result=False):
+        self._single_result = single_result
+        self._result_thread = ResultThread(chunk_object_ids, callback,
+                                           error_callback)
+        self._result_thread.start()
+
+    def wait(self, timeout=None):
+        start = time.time()
+        while timeout is None or time.time() - start < timeout:
+            if self._result_thread.done():
+                break
+            time.sleep(0.001)
+
+    def get(self, timeout=None):
+        self.wait(timeout)
+        if not self._result_thread.done():
+            raise TimeoutError
+
+        results = []
+        for batch in self._result_thread.results():
+            for result in batch:
+                if isinstance(result, PoolTaskError):
+                    raise result.underlying
+            results.extend(batch)
+        if self._single_result:
+            return results[0]
```

Comment:
Not too important, but I prefer combining the `return` with the line below (`... if ... else ...`). Makes it easier to place breakpoints in a debugger, and also to add more logic before the `return` without duplicating code as the code evolves.